Marvelys-Torres/tarea-5/index.js

```diff
@@ -0,0 +1,201 @@
+};
+
+const estaResuelto = (rejilla) => {
+  return (
+    rejilla[0][0] === 1 &&
+    rejilla[0][1] === 2 &&
+    rejilla[0][2] === 3 &&
+    rejilla[0][3] === 4 &&
+    rejilla[1][0] === 5 &&
+    rejilla[1][1] === 6 &&
+    rejilla[1][2] === 7 &&
+    rejilla[1][3] === 8 &&
+    rejilla[2][0] === 9 &&
+    rejilla[2][1] === 10 &&
+    rejilla[2][2] === 11 &&
+    rejilla[2][3] === 12 &&
+    rejilla[3][0] === 13 &&
+    rejilla[3][1] === 14 &&
+    rejilla[3][2] === 15 &&
+    rejilla[3][3] === 0
+  );
+};
+
+const getRejillaAleatoria = () => {
+  let rejilla = [
+    [1, 2, 3, 4],
+    [5, 6, 7, 8],
+    [9, 10, 11, 12],
+    [13, 14, 15, 0],
+  ];
+  let cajaBlanca = new Caja(3, 3);
+  for (let i = 0; i < 1000; i++) {
+    const cajaVecinaAleatoria = cajaBlanca.getCajaVecinaAleatoria();
+    intercambiarCajas(rejilla, cajaBlanca, cajaVecinaAleatoria);
+    cajaBlanca = cajaVecinaAleatoria;
+  }
+  if (estaResuelto(rejilla)) return getRejillaAleatoria();
+  return rejilla;
+};
+
+class Estado {
+  constructor(rejilla, movimiento, tiempo, estado) {
+    this.rejilla = rejilla;
+    this.movimiento = movimiento;
+    this.tiempo = tiempo;
+    this.estado = estado;
+  }
+
+  static listo() {
+    return new Estado(
+      [
+        [0, 0, 0, 0],
+        [0, 0, 0, 0],
+        [0, 0, 0, 0],
+        [0, 0, 0, 0],
+      ],
+      0,
+      0,
+      "listo"
+    );
+  }
+
+  static empezar() {
+    return new Estado(getRejillaAleatoria(), 0, 0, "jugando");
+  }
+}
+
+class Juego {
+  constructor(estado) {
+    this.estado = estado;
+    this.marcaId = null;
+    this.marca = this.marca.bind(this);
+    this.render();
+    this.clickCaja = this.clickCaja.bind(this);
+  }
+
+  static listo() {
+    return new Juego(Estado.listo());
+  }
+
+  marca() {
+    this.setEstado({ tiempo: this.estado.tiempo + 1 });
+  }
+
+  setEstado(nuevoEstado) {
+    this.estado = { ...this.estado, ...nuevoEstado };
+    this.render();
+  }
+
+  clickCaja(caja) {
+    return function () {
+      const cajasVecinas = caja.getCajasVecinas();
+      const cajaBlanca = cajasVecinas.find(
+        (cajaVecina) => this.estado.rejilla[cajaVecina.y][cajaVecina.x] === 0
+      );
+      if (cajaBlanca) {
+        const nuevaRejilla = [...this.estado.rejilla];
+        intercambiarCajas(nuevaRejilla, caja, cajaBlanca);
+        if (estaResuelto(nuevaRejilla)) {
+          clearInterval(this.marcaId);
+          this.setEstado({
+            estado: "gano",
+            rejilla: nuevaRejilla,
+            movimiento: this.estado.movimiento + 1,
+          });
+        } else {
+          this.setEstado({
+            rejilla: nuevaRejilla,
+            movimiento: this.estado.movimiento + 1,
+          });
+        }
+      }
+    }.bind(this);
+  }
+  render() {
+    const { rejilla, movimiento, tiempo, estado } = this.estado;
+    const nuevaRejilla = document.createElement("div");
+    nuevaRejilla.className = "rejilla";
+    for (let i = 0; i < 4; i++) {
```

Comment:
@MarvelysTorres , use modern approach like forEach, map or reduce, also name correctly the variable i